src/main/kotlin/com/emberjs/hbs/HbsHelperCompletionProvider.kt

```diff
@@ -0,0 +1,44 @@
+package com.emberjs.hbs
+
+import com.emberjs.index.EmberNameIndex
+import com.emberjs.lookup.EmberLookupElementBuilder
+import com.emberjs.resolver.EmberName
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionProvider
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.openapi.util.Condition
+import com.intellij.util.CommonProcessors.CollectProcessor
+import com.intellij.util.FilteringProcessor
+import com.intellij.util.ProcessingContext
+import com.intellij.util.indexing.FileBasedIndex
+import com.intellij.util.indexing.FindSymbolParameters.searchScopeFor
+
+/**
+ * The `HbsHelperCompletionProvider` class adds all helper names
+ * from the `EmberNameIndex` to the `CompletionResultSet`.
+ */
+class HbsHelperCompletionProvider : CompletionProvider<CompletionParameters>() {
+
+    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext?, result: CompletionResultSet) {
+        val project = parameters.originalFile.project
+        val scope = searchScopeFor(project, true)
+
+        val collector = CollectProcessor<EmberName>()
+        val filter = FilteringProcessor<EmberName>(Condition { it.type == "helper" }, collector)
+
+        val index = FileBasedIndex.getInstance()
+
+        // Collect all components from the index
+        index.processAllKeys(EmberNameIndex.NAME, filter, scope, null)
+
+        val elements = collector.results
+
+                // Filter out components that are not related to this project
+                .filter { index.getContainingFiles(EmberNameIndex.NAME, it, scope).isNotEmpty() }
```

Comment:
@Turbo87, yes, sure, I've stepped into this couple of times myself.
If you take a look at the processAllKeys [JavaDoc](https://github.com/JetBrains/intellij-community/blob/master/platform/indexing-api/src/com/intellij/util/indexing/FileBasedIndex.java#L143) you'll notice that it guarantees only that keys from the given project will be up-to-date. Actually it returns all keys from the index, including keys that belong to other projects, but they may be not valid. Reason for this is quite simple: performance. Quite often you need keys to do something else, e.g. you provide all names to go to class, but all invalid keys are filtered afterwards via values processing (which returns values only from given scope). If the keys were filtered on the getAllKeys call you would have made same job twice.
This is pretty common mistake, so if you have any ideas on how we could improve JavaDoc to explain that better, it would be awesome